### DISCUSSION TITLE
Have Child widget of Group Resize to fill parent width pixel perfectly

### DISCUSSION THREAD

**Discussion by wyhinton** [fltk-rs/fltk-rs#738] [Q&A]
I'm trying to figure out how to get a child widget to conform to the dimensions of it's parent widget. It basically works, but I have the problem of the updates to the child "lagging behind" the redrawing of the parent. This results in the parent widget being about 1 or 2px wider than the child widget. 
```rust
use fltk::{app, button::Button, prelude::*, window::Window, enums::*, group::Group, frame::Frame, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    wind.make_resizable(true);

    let mut group = Group::new(50,50,300,200,"");
    group.set_frame(FrameType::FlatBox);
    group.set_color(Color::Red);
    group.make_resizable(false);

    let mut left_frame= Frame::new(group.x(),group.y(),50,50,"fill");
    left_frame.set_frame(FrameType::FlatBox);
    left_frame.set_color(Color::Blue);
    left_frame.draw(move|widg|{
      let width =  widg.parent().unwrap().width();;
      widg.resize(widg.x(), widg.y(),width,50);
    });

    group.end();
    let mut frame_left_cl = left_frame.clone();
    group.handle(move|widg, ev|{
      match ev{
        Event::Resize=>{
          frame_left_cl.set_damage_type(Damage::User1);
          widg.set_damage_type(Damage::User1);
        true
        }
        _=>false              
      }
    });
    wind.end();
    wind.show();
    app.run().unwrap();
}
```
![resizing_issue](https://user-images.githubusercontent.com/38958118/122653794-24643f00-d115-11eb-903a-f918c5a84f8b.gif)

I know that ```Pack``` does this kind of resizing, but for reasons I need use this effect for a Group. 

Any thoughts on how to fix this? After taking a look at the [FLTK FL_Pack.cxx](https://github.com/fltk/fltk/blob/master/src/Fl_Pack.cxx), I think the answer might involve setting up a ```draw``` for ```group```, not ```left_frame```, that handles the resizing, but I'm not sure exactly how to go about this.

**Comment by wyhinton** [comment]
Drawing the child widget via the groups draw call still results in about a pixels difference in their size:

```rust
use fltk::{app, button::Button, prelude::*, window::Window, enums::*, group::Group, frame::Frame, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    wind.make_resizable(true);

    let mut group = Group::new(50,50,300,200,"");
    group.make_resizable(false);
    group.set_frame(FrameType::FlatBox);
    group.set_color(Color::Red);

    //hug the left side of the group
    let mut left_frame= Frame::new(group.x(),group.y(),50,50,"fill");
    left_frame.set_frame(FrameType::FlatBox);
    left_frame.set_color(Color::Blue);

    group.end();
    let mut frame_left_cl = left_frame.clone();
    group.draw(move|widg|{
      widg.set_damage(true);
      let mut child = widg.child(0).unwrap();
      child.resize(child.x(), child.y(), widg.width(), child.height());
      child.set_damage(true);
      let mut test = unsafe {widget::Widget::from_widget_ptr(child.as_widget_ptr())};
      widg.draw_child(&mut test);  
      widg.update_child(&mut test);
    });

    wind.end();
    wind.show();
    app.run().unwrap();
}
```
![image](https://user-images.githubusercontent.com/38958118/122654061-cd5f6980-d116-11eb-95d6-0d59dc42ae2e.png)

![image](https://user-images.githubusercontent.com/38958118/122654073-dfd9a300-d116-11eb-811a-04ccb9b7abd3.png)

**Reply by MoAlyousef** [reply]
I’m not able to test things right now, but normally setting damage signals a draw in the next draw cycle, so it might always lag. Can you try calling frame_left_cl.resize() in the Event::Resize of the group handler.

**Reply by wyhinton** [reply]
@MoAlyousef thanks for your response. Yes, calling resize in ```Event::Resize``` did work. 

I guess I was hoping for a solution that only required the groups ```draw()``` function, as doing the resize in the event handle is a bit less compact.

Seems like Fl_Pack.cxx is able to do correct resizing of the children without needing to handle any events, but I can't tell how what I was doing in my first example differs. 

```rust
use fltk::{app, button::Button, prelude::*, window::Window, enums::*, draw::*, group::Group, frame::Frame, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    wind.make_resizable(true);

    //intialize group
    let mut group = Group::new(50,50,300,200,"");
    group.make_resizable(false);
    group.set_frame(FrameType::FlatBox);
    group.set_color(Color::Red);

    //initialize frame
    let mut left_frame= Frame::new(group.x(),group.y(),50,50,"fill");
    left_frame.set_frame(FrameType::FlatBox);
    left_frame.set_color(Color::Blue);

    group.end();
    let mut left_frame_cl = left_frame.clone();
    group.handle(move|widg, ev|{
      match ev{
        Event::Resize=>{
          left_frame_cl.resize(widg.x(), widg.y(), widg.width(), left_frame_cl.height());
          true
        }
        _=>false              
      }
    });
    //remove draw function
    //create a clone the frame in order to move it into draw callback
    // let mut frame_left_cl = left_frame.clone();
    // group.draw(move|widg|{
      //get child 0 (the frame we created)
      // let mut child = widg.child(0).unwrap();
      // draws the child widget correctly on app start
      // child.set_damage(true);
      //create a Widget from our <Box<dyn WidgetExt>>
      // let mut child_as_widget = unsafe {widget::Widget::from_widget_ptr(child.as_widget_ptr())};
      //resize the child
      // child_as_widget.resize(widg.x(), widg.y(), widg.width(), child.height());
      // update the child
      // widg.update_child(&mut child_as_widget);
    // });

    wind.end();
    wind.show();
    app.run().unwrap();
}
```

**Reply by wyhinton** [reply]
[Erco seems to suggest ](https://groups.google.com/g/fltkgeneral/c/UFzrRnaMh2k) that we not use ```resize``` in side of a ```draw()``` call:

> I don't know the language you posted, but it looks like you might be resizing
> in the draw() method, that'd explain why you're seeing a lag.
> 
> Don't do resizing in draw().
> The ONLY thing you should really do in draw() code is draw stuff.
> 
> If you want precise resizing, subclass the group widget and override the resize(x,y,w,h)
> method, and inside that:
> 
>     1) Tell the baseclass to resize(),
>     2) Set the child to the size you want
>     3) Call init_sizes() last, so the group knows you changed the size of one of its children

@MoAlyousef looks like [```init_sizes``` is a function in Fl_Group](https://www.fltk.org/doc-1.3/classFl__Group.html#ad6b50dddf8f0914dc6a7b89ab3663930), but is not yet in ```GroupExt``` ? I do see  a ```init_sizes``` for [various table functions](https://docs.rs/fltk/1.0.19/fltk/prelude/trait.GroupExt.html?search=init_sizes).

**Reply by MoAlyousef** [reply]
Another option is to remove the frame outside of the group, set the window's resizable as the group and set a custom draw callback for the group to resize the frame (this avoids handling things in Event::Resize):
```rust
use fltk::{app, button::Button, prelude::*, window::Window, enums::*, group::Group, frame::Frame, *};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    
    let mut group = Group::new(50,50,300,200,"");
    group.set_frame(FrameType::FlatBox);
    group.set_color(Color::Red);
    group.end();

    //hug the left side of the group
    let mut left_frame= frame::Frame::new(group.x(),group.y(),group.w(),50,"fill");
    left_frame.set_frame(FrameType::FlatBox);
    left_frame.set_color(Color::Blue);
    
    group.draw(move |w| {
        left_frame.resize(w.x(), w.y(), w.w(), 50);
    });

    wind.resizable(&group);

    wind.end();
    wind.show();
    app.run().unwrap();
}
```
